### PR TITLE
[Runtime] Fixed remoting bug where too specific a type reflected.

### PR DESCRIPTION
**Pull request by alexanderkyte**

### CODE REVIEW

**Comment by Therzok** [comment]
:heart:

**Comment by akoeplinger** [comment]
Reference: https://bugzilla.xamarin.com/show_bug.cgi?id=26833

**Comment by tritao** [comment]
Seems like you moved some code around which makes it harder to see what the actual fix was. When you do that it's better to actually make multiple commits, one for the code move and another for the other parts of the changes. Using a visual client like SourceTree makes that really easy to do, you can select the chunks of changes you want in each commit.

**Comment by alexanderkyte** [comment]
@tritao: I was considering splitting the commit but was advised by Kumpera to make it a single commit to allow an easier git bisect. I can easily split the commit if that's more desired.

**Comment by knocte** [comment]
AFAIU if you split the commit in the way @tritao suggests, the first commit would not change behaviour, while the second would (so git-bisect wise, we're good).

**Comment by alexanderkyte** [comment]
I tried to minimize line damage. Tell me if there is anything you think should be moved.

**Comment by knocte** [comment]
Nitpick: if you prefix with [Runtime] one could think that you're changing mono's VM (runtime), but no, you're changing the corlib assembly. Therefore both commits should be prefixed with [corlib] or [remoting].

**Comment by tritao** [comment]
We really could use with a test case here if it's not too hard to isolate the problem. Fix itself looks good to me thought I'm not very familiar with this part of the codebase...

**Comment by alexanderkyte** [comment]
I'm not entirely familiar with the way that xunit had classes loaded on one appdomain but not the other. I'll see how difficult it would be to recreate such a situation.

**Comment by alexanderkyte** [comment]
This bug only shows it's head in a very specific situation that I'm having trouble duplicating. You need to have a CrossAppDomainChannel open, so that it serializes/deserializes the data using the given CADMessage helpers. It needs to be a MarshalByRef object. And It can't be a TransparentProxy.

File: mcs/class/corlib/System.Runtime.Remoting.Messaging/CADMessages.cs [hunk 1]
```diff
@@ -74,7 +74,88 @@ internal class CADMessageBase {
...
+		internal RuntimeMethodHandle MethodHandle;
+		internal string FullTypeName;
+		internal MethodBase _method;
+
+		public CADMessageBase (IMethodMessage msg) {
+			MethodHandle = msg.MethodBase.MethodHandle;
+			FullTypeName = msg.MethodBase.DeclaringType.AssemblyQualifiedName;
+		}
+
+		internal MethodBase method {
```
Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Messaging/CADMessages.cs]: Properties must start with a capital letter.

File: mcs/class/corlib/System.Runtime.Remoting.Messaging/CADMessages.cs [hunk 2]
```diff
@@ -74,7 +74,88 @@ internal class CADMessageBase {
...
+			get {
+				if (_method == null) {
+					_method = GetMethod();
+				}
+				return _method;
+			}
+		}
+
+		internal MethodBase GetMethod ()
```
Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Messaging/CADMessages.cs]: This big method was moved. I agree with the others, having the move in a separate commit would make it clear if anything changed in it. (did it change?)

Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Messaging/CADMessages.cs]: you did it already, sorry for the confusion.

**Comment by kumpera** [comment]
PR looks almost ok. What's missing is a better commit message for the second commit that explains what exactly happened before and what the fix does.

An illustrative example of how the old code failed would go great lengths in making this ready.

The commit message body must answer the questions the headline brings.

Why too specific was a problem?
What sort of type combination are an issue?
Where was this seem?
...
